core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -149,29 +162,70 @@ class BlockManagerMasterEndpoint(
     // First remove the metadata for the given RDD, and then asynchronously remove the blocks
     // from the slaves.
 
+    // The message sent to the slaves to remove the RDD
+    val removeMsg = RemoveRdd(rddId)
+
     // Find all blocks for the given RDD, remove the block from both blockLocations and
-    // the blockManagerInfo that is tracking the blocks.
+    // the blockManagerInfo that is tracking the blocks and create the futures which asynchronously
+    // remove the blocks from slaves and gives back the number of removed blocks
     val blocks = blockLocations.asScala.keys.flatMap(_.asRDDId).filter(_.rddId == rddId)
+    val blocksToDeleteByShuffleService =
+      new mutable.HashMap[BlockManagerId, mutable.HashSet[RDDBlockId]]
+
     blocks.foreach { blockId =>
-      val bms: mutable.HashSet[BlockManagerId] = blockLocations.get(blockId)
-      bms.foreach(bm => blockManagerInfo.get(bm).foreach(_.removeBlock(blockId)))
-      blockLocations.remove(blockId)
+      val bms: mutable.HashSet[BlockManagerId] = blockLocations.remove(blockId)
+
+      val (bmIdExtShuffle, bmIdExecutor) = bms.partition(_.port == externalShuffleServicePort)
+      if (bmIdExecutor.isEmpty && bmIdExtShuffle.nonEmpty) {
```

Comment:
Is this correct? What happens when you have a block with replication=2, for example, with one block in a live executor and one in a dead one?

As far as I can tell you'd fail to remove the blocks from the dead executor's local dirs in that case.

Reply:
Very good catch I have missed this. I can extend the `blockStatusByShuffleService` with a flag (stored for each blockId) indicating the cleanup must be done via the shuffle service and update this info when the executor is removed.

This seams to me the right time to extract `JHashMap[BlockId, BlockStatus]` from 

~~~scala
private val blockStatusByShuffleService =
    new mutable.HashMap[BlockManagerId, JHashMap[BlockId, BlockStatus]]
~~~

Into a new class: like `ShuffleServiceBlockInfo`.

Reply:
Actually there is a smaller fix for this problem: the cleanup should be done via Shuffle Service if there is no live executor for it.